### ISSUE TITLE
Tests with long-double precision fail on some architectures

### ISSUE THREAD

**Issue by ghisvail** [h5py/h5py#817]
In addition to the test failures reported in #816, builds on architectures which do not support long-double precision (such as ppc64el) produce the following output:
```
======================================================================
FAIL: test_complex256 (h5py.tests.old.test_dataset.TestCreateShape)
Confirm that the default dtype is float
----------------------------------------------------------------------
Traceback (most recent call last):
  File "h5py/tests/old/test_dataset.py", line 101, in test_complex256
    self.assertEqual(dset.dtype, np.dtype('complex256'))
AssertionError: dtype('complex128') != dtype('complex256')

======================================================================
FAIL: test_long_double (h5py.tests.old.test_dataset.TestCreateShape)
Confirm that the default dtype is float
----------------------------------------------------------------------
Traceback (most recent call last):
  File "h5py/tests/old/test_dataset.py", line 94, in test_long_double
    self.assertEqual(dset.dtype, np.longdouble)
AssertionError: dtype('float64') != <type 'numpy.float128'>
```

**Comment by tacaswell** [comment]
Fails in what looks like a related way in powerpc

```
======================================================================
ERROR: test_complex256 (h5py.tests.old.test_dataset.TestCreateShape)
Confirm that the default dtype is float
----------------------------------------------------------------------
Traceback (most recent call last):
  File "h5py/tests/old/test_dataset.py", line 101, in test_complex256
    self.assertEqual(dset.dtype, np.dtype('complex256'))
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/«PKGBUILDDIR»/h5py/_objects.c:2682)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/«PKGBUILDDIR»/h5py/_objects.c:2640)
  File "h5py/_hl/dataset.py", line 238, in dtype
    return self.id.dtype
  File "h5py/h5d.pyx", line 120, in h5py.h5d.DatasetID.dtype.__get__ (/«PKGBUILDDIR»/h5py/h5d.c:2376)
  File "h5py/h5d.pyx", line 123, in h5py.h5d.DatasetID.dtype.__get__ (/«PKGBUILDDIR»/h5py/h5d.c:2310)
  File "h5py/h5t.pyx", line 358, in h5py.h5t.TypeID.dtype.__get__ (/«PKGBUILDDIR»/h5py/h5t.c:4886)
  File "h5py/h5t.pyx", line 359, in h5py.h5t.TypeID.dtype.__get__ (/«PKGBUILDDIR»/h5py/h5t.c:4843)
  File "h5py/h5t.pyx", line 1103, in h5py.h5t.TypeCompoundID.py_dtype (/«PKGBUILDDIR»/h5py/h5t.c:11103)
  File "h5py/h5t.pyx", line 966, in h5py.h5t.TypeFloatID.py_dtype (/«PKGBUILDDIR»/h5py/h5t.c:9958)
ValueError: Insufficient precision in available types to represent (127, 116, 11, 0, 116)

======================================================================
ERROR: test_long_double (h5py.tests.old.test_dataset.TestCreateShape)
Confirm that the default dtype is float
----------------------------------------------------------------------
Traceback (most recent call last):
  File "h5py/tests/old/test_dataset.py", line 94, in test_long_double
    self.assertEqual(dset.dtype, np.longdouble)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/«PKGBUILDDIR»/h5py/_objects.c:2682)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/«PKGBUILDDIR»/h5py/_objects.c:2640)
  File "h5py/_hl/dataset.py", line 238, in dtype
    return self.id.dtype
  File "h5py/h5d.pyx", line 120, in h5py.h5d.DatasetID.dtype.__get__ (/«PKGBUILDDIR»/h5py/h5d.c:2376)
  File "h5py/h5d.pyx", line 123, in h5py.h5d.DatasetID.dtype.__get__ (/«PKGBUILDDIR»/h5py/h5d.c:2310)
  File "h5py/h5t.pyx", line 358, in h5py.h5t.TypeID.dtype.__get__ (/«PKGBUILDDIR»/h5py/h5t.c:4886)
  File "h5py/h5t.pyx", line 359, in h5py.h5t.TypeID.dtype.__get__ (/«PKGBUILDDIR»/h5py/h5t.c:4843)
  File "h5py/h5t.pyx", line 966, in h5py.h5t.TypeFloatID.py_dtype (/«PKGBUILDDIR»/h5py/h5t.c:9958)
ValueError: Insufficient precision in available types to represent (127, 116, 11, 0, 116)

======================================================================
```

**Comment by tacaswell** [comment]
To quote wikipedia https://en.wikipedia.org/wiki/Long_double

> On some PowerPC and SPARCv9 machines,[citation needed] long double is implemented as a double-double arithmetic, where a long double value is regarded as the exact sum of two double-precision values, giving at least a 106-bit precision; with such a format, the long double type does not conform to the IEEE floating-point standard. Otherwise, long double is simply a synonym for double (double precision).

so it looks like these features are just un-supported on those arch.  The ones that error out seem to be behaving in a reasonable way (it will tell you loudly what you want is not supported).

**Comment by ghisvail** [comment]
I added an error message to the float promotion test for long-double. This what you get on `ppc64el`:
```
======================================================================
FAIL: test_custom_float_promotion (h5py.tests.old.test_h5t.TestTypeFloatID)
Custom floats are correctly promoted to standard floats on read.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "h5py/tests/old/test_h5t.py", line 209, in test_custom_float_promotion
    msg="{} != {}".format(dset.dtype, np.longdouble))
AssertionError: float64 != <type 'numpy.float128'>
```

**Comment by tacaswell** [comment]
The ppc64 failures like like they may be due to how the datatypes are defined.  We construct the complex data dtypes in h5t.pyx it `_c_complex` with

```python
  elif length == 32:
        IF COMPLEX256_SUPPORT:
            size = h5py_size_n256
            off_r = h5py_offset_n256_real
            off_i = h5py_offset_n256_imag
            tid_sub = H5T_NATIVE_LDOUBLE
        ELSE:
            raise TypeError("Illegal length %d for complex dtype" % length)
```

The conditional comes from a check if `hasattr(numpy, 'complex256')`, but if `H5T_NATIVE_LDOUBLE` is really a normal (64 bit) double the h5t object we hold for `complex256` is really a `complex128`, which h5py then correctly write to disk and when we read it back out has the wrong type. 

@andreabedini found this wlle digging http://hdf-forum.184993.n3.nabble.com/Issues-with-H5T-NATIVE-LDOUBLE-td4026450.html ;)

**Comment by andreabedini** [comment]
@tacaswell :-)

I haven't been following and I am not sure this is relevant here but working on the long double support for PyTables I came to the conclusion that one should avoid referring to 128 or 256 version of types where possible since they are platform dependent. At C level the correct type is long double which numpy exposes with `numpy.longdouble` and  `numpy.clongdouble`.

From what you write above it looks like the HDF5 library thinks that on ppc64 `sizeof(long double)` is 12 while numpy thinks it's 8. 🤔

**Comment by mraspaud** [comment]
I have to concur with @andreabedini : sometimes (on 32bits systems), a longdouble is 96 bits...

**Comment by tacaswell** [comment]
The `complex256` datatype was added to support the pytables rebase, not sure what to do about this....

**Comment by andreabedini** [comment]
@tacaswell true, I did make this point in https://github.com/h5py/h5py/pull/749#issuecomment-246216978 but it might be considered my fault not thinking about it before merging #749 😞

**Comment by tacaswell** [comment]
@AdamWill Can I nerd-snipe you into looking at more arch-dependent integer length issues?

These are arising from debian builds, but fedora is going to run into these soon as well :wink: .

**Comment by tacaswell** [comment]
https://buildd.debian.org/status/logs.php?pkg=h5py&ver=2.7.0~rc2-2 <- new logs.

It looks like on powerpc and ppc64 we are getting tests that fail due to the system (correctly?) deciding that it does not have enough bits.  The ppc64el one is more worrying as there is some miss-match between the names, the actually bit depth, and what numpy thinks the bit depth is.

I would argue that only the ppc64el one should be a blocking issue (as the other ones could be dealt with by known-fails).

**Comment by tacaswell** [comment]
I am starting to suspect that this may be a numpy and/or hdf5 bug.  The float conversion test does this:

```python
dset = h5py.h5d.create(fid, dataset5, h5t.NATIVE_LDOUBLE, space)
dset.write(h5py.h5s.ALL, h5py.h5s.ALL, wdata2)
dset = f[dataset5]
self.assert_(dset.dtype == np.longdouble)
```

where `h5t.NATIVE_LDOUBLE` is defined in h5t.pyx as 

```python
NATIVE_LDOUBLE = lockid(H5T_NATIVE_LDOUBLE)
```

and H5T_NATIVE_LDOUBLE comes from `api_types_hdf5` which means it is extracted from hdf5 it's self.  

https://docs.scipy.org/doc/numpy-dev/user/basics.types.html#extended-precision documents that `np.longdouble` should be the native `long double` type.  

https://support.hdfgroup.org/HDF5/doc/RM/PredefDTypes.html documents that `H5T_NATIVE_LDOUBLE` should also be the native `long_double` 

Clearly one of the two is incorrectly identifying the native size of a 'long double' (I don't know if the `'_'` vs `' '` in the names is important, I hope not).

I still need to chase through the numpy dtype -> h5t dtype conversion code, but my gut guess is that h5py is assuming that `np.longdouble`  is interchangeable with `H5T_NATIVE_LDOUBLE` which the example above demonstrates.

**Comment by tacaswell** [comment]
For reference, these are the three concerning failures:

```

======================================================================
FAIL: test_complex256 (h5py.tests.old.test_dataset.TestCreateShape)
Confirm that the default dtype is float
----------------------------------------------------------------------
Traceback (most recent call last):
  File "h5py/tests/old/test_dataset.py", line 101, in test_complex256
    self.assertEqual(dset.dtype, np.dtype('complex256'))
AssertionError: dtype('complex128') != dtype('complex256')

======================================================================
FAIL: test_long_double (h5py.tests.old.test_dataset.TestCreateShape)
Confirm that the default dtype is float
----------------------------------------------------------------------
Traceback (most recent call last):
  File "h5py/tests/old/test_dataset.py", line 94, in test_long_double
    self.assertEqual(dset.dtype, np.longdouble)
AssertionError: dtype('float64') != <type 'numpy.float128'>

======================================================================
FAIL: test_custom_float_promotion (h5py.tests.old.test_h5t.TestTypeFloatID)
Custom floats are correctly promoted to standard floats on read.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "h5py/tests/old/test_h5t.py", line 208, in test_custom_float_promotion
    self.assertEqual(dset.dtype, np.longdouble)
AssertionError: dtype('float64') != <type 'numpy.float128'>

----------------------------------------------------------------------
```

**Comment by tacaswell** [comment]
I retract my bug theory above, I have a better theory:

We now walk up the bit depth chain to find the smallest float which will fit the encoded data type.  On ppc64 numpy provides a `np.float128` and `np.complex256` which are in fact the same size as `np.float64` and `np.complex128` due to `long double` being available in the compiler, but being the same as a `double` (this part is a bit of wild speculation).   Thus, data is round tripping correctly, but our test is pinned to the higher one.

The fix might be as simple as adjusting the tests to check if `np.float64` and `np.float128` are really the same size and skipping or adjusting as desired.

**Comment by tacaswell** [comment]
https://gcc.gnu.org/onlinedocs/gcc/Floating-Types.html seems to be the gcc docs for what extended types to expect which seems to say that some powerpc 64 machines do have '128' bit types, but I do not know enough to be user that applies to this case or not.

I am now hoping that my stream of research is enough for someone who knows more to go "oh, the problem is obviously....." ;)

**Comment by ghisvail** [comment]
There is also the section on Numpy's [extended precision types](https://docs.scipy.org/doc/numpy/user/basics.types.html#extended-precision). In particular:
> In spite of the names, np.float96 and np.float128 provide only as much precision as np.longdouble, that is, 80 bits on most x86 machines and 64 bits in standard Windows builds.

**Comment by tacaswell** [comment]
attn @epourmal Is this something we should bring to hdf5 upstream?

**Comment by mraspaud** [comment]
@tacaswell Your last theory sounds interesting, but shouldn't equality work then ? If `np.float128` really is `np.float64`, shouldn't `np.float64 == np.float128` be true ?
@ghisvail if you have the possibility, could paste the `np.finfo` of  `np.float128` and `np.float64` on ppc64le ? Especially the mantissa and exponent size, just to check how they actually different.

**Comment by mraspaud** [comment]
Ok, I have a new theory: 
as wikipedia says, on ppc64le `long double` is probably a `double double`, so the mantissa size should be 106 bits, but HDF5 reports only 52 bits (see the inline comments in this patch https://bugzilla.redhat.com/show_bug.cgi?id=1078173 )
So if we also check the total size of the type, we should be able to detect this case. I'm making a PR.

**Comment by andreabedini** [comment]
I would hate to see [this](http://hdf-forum.184993.n3.nabble.com/Issues-with-H5T-NATIVE-LDOUBLE-td4026450.html) hitting again (long thread, scroll down).

tl;dr Few years ago I found that HDF5 is relying on undefined behaviour (in `H5detect.c`, which is run at `./configure` time) to find out the float bit representation on different architectures. Of course compilers ended up breaking that code and long double was detected with the wrong size. At that time HDFGroup just replaced the undefined behaviour with some other undefined behaviour. RedHat bug report above seems to compatible. 

@mraspaud at that time (for PyTables) we had to disable long doubles on the broken architectures (with a runtime check). See [here](https://github.com/PyTables/PyTables/blob/44265c6b3f10ba4dd0d528ea54636c1fe05d6353/tables/__init__.py#L192).
